File: src/algorithms/cpp/Patching_Array.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


class Solution {
public:
    int minPatches(vector<int>& nums, int n) {
        long long upper_bound = 0;
        int i = 0, result = 0;
        while (upper_bound < n) {
            if (i < SZ(nums) && upper_bound + 1 >= nums[i]) {
                upper_bound += nums[i ++];
            }
            else {
                result += 1;
                upper_bound = 2 * upper_bound + 1;
            }
        }
        return result;
    }
};


int main() {
    return 0;
}

File: src/algorithms/python/Patching_Array.py
#!/bin/env python2
# -*- coding: utf-8 -*-


class Solution(object):
    def minPatches(self, nums, n):
        """
        :type nums: List[int]
        :type n: int
        :rtype: int
        """
        upper_bound, i, result = 0, 0, 0
        while upper_bound < n:
            if i < len(nums) and upper_bound + 1 >= nums[i]:
                upper_bound, i = upper_bound + nums[i], i + 1
            else:
                result, upper_bound = result + 1, 2 * upper_bound + 1
        return result
